### ISSUE TITLE
Union Fails

### ISSUE THREAD

**Issue by srt19170** [w8r/martinez#51]
Performing a union on the following two polygons causes an error:

```
Uncaught TypeError: Cannot read property 'left' of null
    at rotateLeft (martinez.js:123)
    at AVLTree.insert (martinez.js:572)
    at subdivide (martinez.js:1733)
    at boolean (martinez.js:1379)
    at Object.12.boolean.union (martinez.js:1389)
```
Polygons (I did not try to simplify these): 

```
	let poly1 = [[0, 0],
		     [10.502137906098726, -8.91955472410577],
		     [19.362154173294872, -16.536565170635974],
		     [26.806688712234628, -23.184332654075142],
		     [33.06238143356417, -29.196158488907805],
		     [38.35587224792968, -34.9053439896185],
		     [42.91380106597733, -40.64519047069176],
		     [49.10713251970216, -50.62072053266144],
		     [46.96280779835334, -46.74899924661212],
		     [50.72953235570385, -53.55007163186412],
		     [54.440614648675066, -61.381708940932285],
		     [58.32269458791316, -70.57721248830114],
		     [62.602412084064326, -81.46988358845522],
		     [67.50640704777474, -94.39302355587905],
		     [68.81850151153644, -97.30806695169885],
		     [69.8621573393369, -97.95163015963242],
		     [71.06094548688412, -96.8158900656893],
		     [76.97887880292775, -92.44160559185252],
		     [77.08733590764791, -82.61594306225551],
		     [76.04410744893605, -80.24754616339085],
		     [77.94689401460623, -84.56736102628204],
		     [86.1143714390989, -73.17583405575341],
		     [92.5064384837837, -63.06061283731647],
		     [97.5579645467036, -54.03110246393672],
		     [101.70381902590161, -45.89670802857968],
		     [105.37887131942071, -38.466834624210875],
		     [111.17983390680583, -28.31159335439475],
		     [109.0179908253039, -31.55088734379581],
		     [116.19929795079966, -20.790455559126197],
		     [116.05604694159416, -24.9582712803],
		     [120.92790906633451, -18.49839152668898],
		     [127.06844659756788, -11.980653175928243],
		     [134.91252893333734, -5.214461320983322],
		     [144.89502547168587, 0],
		     [0, 0]];

	let poly2 = [[-42.061875001865225, 0],
		     [-38.333456448806956, -4.191434705875763],
		     [-35.788781183126574, -3.2898181160071496],
		     [-33.479009769666234, -11.900923266495782],
		     [-30.45530277326811, -18.141137813496414],
		     [-28.381278877922217, -16.789914833141882],
		     [-24.233743423095476, -10.47911691415063],
		     [-20.015414375728398, -17.01009233689141],
		     [-17.729009702761488, -18.56589936873845],
		     [-11.159466795267058, -11.53852312789438],
		     [-6.31966948099911, -5.434229624518701],
		     [-1.03024624278126, -3.200354520901266],
		     [6.888174436562863, 0],
		     [-42.061875001865225, 0]];
```
However, intersection works.

**Comment by rowanwins** [comment]
Thanks for your report @srt19170 . poly1 appears to be a pretty strange geometry with self-intersections so there must be something in that that is throwing it out...

**Comment by w8r** [comment]
Quite tired of these tree exceptions. From my recent experience, [splay-tree](https://github.com/w8r/splay-tree) turns out MUCH more stable with less predictable input. Performance is pretty much the same. I will run a test when looking into this issue.

**Comment by srt19170** [comment]
Doing some fuzz testing, it's easy to find examples that cause a stack error when connectEdges gets struck.  Example:

```
p1 = [[76.82,43.37],[94.82,58.46],[8.76,50.75],[40.62,37.06],[74.37,28.70],[2.26,78.49],[25.69,32.82],[43.28,99.31],[89.84,1.99],[76.82,43.37]];
p2 = [[57.29,65.63],[87.12,88.37],[34.01,45.05],[62.99,46.26],[65.62,6.44],[84.48,13.87],[5.99,9.30],[97.01,83.87],[24.18,60.79],[57.29,65.63]];
```
I had to add some code to connectEdges to throw an exception when contour[0] is growing out of control so I could catch these errors.

Other errors:
```
TypeError: Cannot read property 'left' of null
    at rotateLeft (martinez.js:123)
    at AVLTree.remove (martinez.js:671)
    at subdivide (martinez.js:1774)
    at boolean (martinez.js:1381)
    at Object.boolean.union (martinez.js:1391)

p1 = [[91.86,71.89],[4.14,39.29],[47.18,71.33],[59.14,88.86],[45.95,51.60],[10.55,91.54],[87.30,56.67],[35.38,6.75],[69.84,76.58],[91.86,71.89]];
p2 = [[59.25,56.44],[43.71,27.89],[58.99,93.35],[88.30,15.10],[21.64,67.85],[7.98,10.73],[59.32,20.05],[68.66,43.19],[2.84,52.94],[59.25,56.44]];
```
and 

```
TypeError: Cannot read property 'point' of undefined
    at nextPos (martinez.js:1066)
    at connectEdges (martinez.js:1130)
    at boolean (martinez.js:1385)
    at Object.boolean.intersection (martinez.js:1406)

p1 = [[89.06,52.71],[75.83,17.71],[99.01,68.48],[5.65,89.14],[28.31,50.57],[56.47,67.18],[99.24,94.94],[13.49,54.72],[2.88,45.70],[89.06,52.71]];
p2 = [[78.95,36.44],[17.01,87.75],[29.98,13.36],[21.28,19.81],[41.98,44.88],[39.68,19.80],[29.41,81.88],[58.12,4.71],[10.07,10.49],[78.95,36.44]];
```
and

```
TypeError: Cannot read property 'balanceFactor' of null
    at AVLTree.remove (martinez.js:671)
    at subdivide (martinez.js:1774)
    at boolean (martinez.js:1381)
    at Object.boolean.union (martinez.js:1391)

p1 = [[2.43,91.94],[17.97,83.55],[51.27,68.83],[50.36,39.18],[93.98,3.38],[71.27,47.89],[69.81,92.08],[94.96,61.57],[3.32,61.57],[2.43,91.94]];
p2 = [[97.48,8.70],[31.31,50.79],[40.10,86.20],[92.23,13.40],[51.26,40.50],[32.02,49.85],[60.16,27.78],[95.69,72.85],[9.54,0.99],[97.48,8.70]];
```
and (only saw this one with longer polygons)

```
TypeError: Cannot read property 'right' of null
    at rotateRight (martinez.js:154)
    at AVLTree.remove (martinez.js:662)
    at subdivide (martinez.js:1774)
    at boolean (martinez.js:1381)
    at Object.boolean.intersection (martinez.js:1406)

p1 = [[19.24,15.92],[61.95,96.79],[65.20,2.87],[74.10,83.68],[86.91,22.43],[21.51,83.87],[86.05,31.98],[53.74,18.77],[39.29,22.37],[43.10,61.44],[13.41,21.18],[74.93,62.02],[33.54,43.49],[65.68,53.65],[82.26,25.67],[24.83,44.94],[70.54,10.43],[86.85,93.46],[97.50,28.08],[19.24,15.92]];
p2 = [[86.80,37.60],[42.57,84.18],[34.79,74.82],[24.37,77.14],[78.49,72.49],[82.72,90.56],[78.11,88.33],[39.79,41.82],[0.46,54.90],[47.93,71.49],[72.92,7.78],[84.25,64.49],[49.33,61.59],[50.84,21.45],[16.46,80.60],[99.74,78.41],[24.10,89.66],[79.23,11.97],[17.05,13.66],[86.80,37.60]];
```
Here's the fuzz testing code I used:

```
	const polysize = 20;
	let p1 = Array(polysize), p2 = Array(polysize);
	let p1str, p2str;
	const exceptions = {};
	for(let i=0;i<1000000;i++) {
	    for(let j=0;j<polysize-1;j++) {
		p1[j] = [Utils.rand(0,100), Utils.rand(0,100)];
		p2[j] = [Utils.rand(0,100), Utils.rand(0,100)];
	    };
	    p1[polysize-1] = p1[0];
	    p2[polysize-1] = p2[0];
	    try {
		let intersect = Martinez.intersection([p1], [p2]);
		let union = Martinez.union([p1], [p2]);
	    } catch (e) {
		if (!(e.toString() in exceptions)) {
		    console.log('Caught exception: '+e);
		    exceptions[e] = e;
		    if (e.stack) console.log(e.stack);
		    p1str = '';
		    for(let j=0;j<p1.length-1;j++) {
			p1str += '['+p1[j][0].toFixed(2)+','+p1[j][1].toFixed(2)+'],';
		    };
		    p1str += '['+p1[p1.length-1][0].toFixed(2)+','+p1[p1.length-1][1].toFixed(2)+']';
		    p2str = '';
		    for(let j=0;j<p2.length-1;j++) {
			p2str += '['+p2[j][0].toFixed(2)+','+p2[j][1].toFixed(2)+'],';
		    };
		    p2str += '['+p2[p2.length-1][0].toFixed(2)+','+p2[p2.length-1][1].toFixed(2)+']';
		    console.log('p1 = ['+p1str+'];');
		    console.log('p2 = ['+p2str+'];');
		};
	    };

	};
	console.log('Completed fuzz test.');
```

**Comment by w8r** [comment]
Case from the first massage is solved: https://codepen.io/w8r/pen/qYezEX

**Comment by w8r** [comment]
one of the other cases works as well https://codepen.io/w8r/pen/KROjdK

**Comment by srt19170** [comment]
Great progress, thanks!

**Comment by w8r** [comment]
all the cases from the last comment are working https://codepen.io/w8r/pen/YLmoyL

**Comment by w8r** [comment]
solved in `v0.4.1`
